### PR TITLE
haskell: Split boot libraries into individual outputs (take 2)

### PR DESCRIPTION
**Pull request by kirelagin**
This is like #43017 (fix for #42069), but:

1. Rebased onto master (now with GHC 8.6)
2. Does only the bare minimum, only moves packages to individual outputs.

I am still not sure what to do with ghcjs (and whether anything needs to be done at all).

@Ericson2314

### CODE REVIEW

**Comment by Ericson2314** [comment]
I saw that after posting here, sorry, and just did a bit there.

**Comment by ElvishJerricco** [comment]
Am I misunderstanding, or isn't this just going to make the issues with too many command line arguments even worse?

**Comment by kirelagin** [comment]
I don’t know 🤷‍♂️.

In any case, this change is orthogonal to those issues and even if it makes them worse, let’s hope they will get eventually resolved by fixing the root cause.

**Comment by Ericson2314** [comment]
@ElvishJerricco It should just add a few more.

**Review by peti — COMMENTED** [review]
Wouldn't it be *far* simpler to clean out all undesirable package config files from the generated `package.conf.d` directory so that only those mentioned in `setup-depends` actually show up in there?

**Review by peti — CHANGES_REQUESTED** [review]
ghc is the wrong place to fix https://github.com/NixOS/nixpkgs/issues/42069. The issue is caused by incorrect code in the generic builder and it ought to be fixed *there*, not in the compiler.

**Comment by kirelagin** [comment]
@peti All this story is not specific to `Setup.hs` and #42069 is essentially only about a regression introduced recently, but the issue is still present not only when building `Setup.hs`. The only reason Haskell packages can be compiled without this change is that `ghc` is being treated as a Haskell dependency; if we fix the generic builder to actually only add Haskell dependencies to pkgdb, none of the boot packages will get there because, as you know, they are `null`-ed out.

**Comment by peti** [comment]
> If we fix the generic builder to actually only add Haskell dependencies to pkgdb, none of the boot packages will get there because, as you know, they are null-ed out.

Yes, this is true. From that point of view, an ideal solution would be to have a derivation per core library and to adapt the overrides to point to them (rather than `null`).

**Comment by peti** [comment]
How did you test this PR?

File: pkgs/development/haskell-modules/make-package-set.nix [hunk 1]
```diff
@@ -148,7 +158,9 @@ let
       '';
     });
 
-in package-set { inherit pkgs stdenv callPackage; } self // {
+in genAttrs (ghc.passthru.bootPackages or []) (name: ghc."${haskellLib.toOutputName name}") //
```
Comment (matthewbauer) [inline pkgs/development/haskell-modules/make-package-set.nix]: You should leave off the ```passthru``` part. It should just be ghc.bootPackages.

**Comment by kirelagin** [comment]
`ghci` works but your question made me think that this is actually a problem, because it is not supposed to. What I should do is regenerate the pkgdb cache after moving the config files away.

To improve UX with ghci we should probably make all ghc environments that are meant to be used interactively include `base` by default.

**Comment by peti** [comment]
> I build a couple of heavy packages (pandoc, stack) ...

What about the 'ghcWithPackages' and 'ghcWithHoogle' functions? Do these
still work and do they produce correctly working environments? Also,
does 'callHackage' still work?

**Comment by kirelagin** [comment]
@peti
`callHackage` works.
`ghcWithPackages` does not. The problem is that dependency propagation doesn’t work for boot packages, e.g. if a package I am adding depends on `binary`, `binary` will be propagated, but `binary` depends on `containers` and containers will not be, because we don’t know anything about dependencies between boot packages.

I think it might be time we stop pulling boot packages from ghc build and instead just grab the matching version from Hackage and compile it again... The number of boot packages that cannot be built this way is hopefully small and we’ll have to handle them separately. Any ideas?

**Comment by Ericson2314** [comment]
@kirelagin I'm all for compiling again. But if need be, we can fake `nix-support/propagated-build-inputs` so things work anyays.

**Comment by kirelagin** [comment]
FTR here is another bug:
`rts` is very special in that it doesn’t have a version so its config file `rts.conf` (without `-version`) doesn’t match the regex I currently use, so it doesn’t get copied and the `rts` output ends up being empty which, in turn, means that packages that depend on it won’t build as `package.conf.d/*.conf` used to populate pkgdb matches no files and thus fails.
Looks like `rts` needs to be special-cased.

**Comment by kirelagin** [comment]
FTR I have just realised that GHC executable itself works only because of the stale pkgdb cache :/.

**Comment by kirelagin** [comment]
I think we should just leave those libraries that are built together with GHC alone as much as possible, just pretend that they are merely a part of the resulting compiler and don’t use them for anything else.

Then we’ll have real derivations for boot libraries, most of them will be just plain versions form Hackage and those that cannot be compiled will be like in #43300.

**Comment by Ericson2314** [comment]
@kirelagin I think that should move us away from a multiple output direction then? If you notice there's some wrapping with a `-B` that is how GHC finds itself. We can replace that in the GHC shim derivation with a different `-B ...` to a boot directory without any libraries.

We can also protype the shimming + fresh separately-built boot library with the prebuilt GHCs and prebuilt-GHC-built-`hscolour, which will help with debug cycles too.

**Comment by mmahut** [comment]
Any updates on this pull request?

**Comment by cdepillabout** [comment]
Similar to https://github.com/NixOS/nixpkgs/pull/39735#issuecomment-623864802, this is quite old, and hasn't seen any updates recently, so I am going to go ahead and close it.

If anyone is interested in pursuing this, please feel free to open a new PR on the current `haskell-updates` branch.

**Pull request closed without merging**